[Feat] 공통 응답 및 전역 예외 처리 구현

jwdebbie opened the pull request:
## 🔍 변경 유형
<!-- 해당하는 모든 항목에 x 표시해주세요 -->
- [x] ✨ 새로운 기능 (Feature)
- [ ] 🐛 버그 수정 (Bug Fix)
- [ ] 📚 문서 업데이트 (Documentation)
- [ ] 🎨 코드 스타일 변경 (Style)
- [ ] ♻️ 코드 리팩토링 (Refactoring)
- [ ] 🧪 테스트 추가/수정 (Tests)
- [ ] 🔧 설정 변경 (Configuration)
- [ ] 🚀 성능 개선 (Performance)
- [ ] 🔒 보안 이슈 (Security)
- [ ] 기타:

## 🔗 관련 이슈
<!-- 연결된 이슈를 작성해주세요. 예: "Closes #123, Fixes #456" -->
Closes #3 

## 📋 변경 내용
<!-- 이 PR에서 변경한 내용을 자세히 설명해주세요 -->
- 공통 응답 포맷 `ApiResponse<T>` 구현
- 전역 예외 처리기 `GlobalExceptionHandler` 구현
- 커스텀 예외 구조 `CustomException`, `ErrorCode` 설계 및 연동

## 📸 스크린샷 / 데모
<!-- UI 변경이나 기능 추가의 경우 스크린샷이나 GIF를 첨부해주세요 -->

## ✅ PR 체크리스트
<!-- PR을 제출하기 전에 아래 체크리스트를 확인해주세요 -->
- [x] 코드 컨벤션을 준수했습니다
- [x] 자체 코드 리뷰를 수행했습니다
- [x] 관련 문서를 업데이트했습니다
- [x] 코드 중복이 없습니다
- [x] 불필요한 주석과 로그를 제거했습니다
- [x] 변수와 함수명이 목적에 맞게 명명되었습니다

## 📌 특별히 리뷰어가 확인해야 할 부분
<!-- 리뷰어에게 특별히 살펴봐야 할 부분이 있다면 언급해주세요 -->

## 💬 추가 정보
<!-- 기타 참고 사항이나 관련 정보가 있다면 적어주세요 -->

jwdebbie committed:
✨ feat: 공통 응답 및 전역 예외 처리 구현 (#3)

supsup-hae commented on src/main/java/cotato/backend/global/error/ErrorCode.java:
```diff
@@ -0,0 +1,21 @@
+    INVALID_INPUT_VALUE("잘못된 입력입니다.", HttpStatus.BAD_REQUEST),
+    NOT_FOUND("존재하지 않는 리소스입니다.", HttpStatus.NOT_FOUND),
+    INTERNAL_SERVER_ERROR("서버 오류가 발생했습니다.", HttpStatus.INTERNAL_SERVER_ERROR),
+    EXTERNAL_API_FAIL("외부 API 호출 실패", HttpStatus.BAD_GATEWAY);
```
미리 프로젝트에서 예상되는 에러코드를 작성해주셨네요 감사합니다 👏🏻

supsup-hae commented on src/main/java/cotato/backend/global/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
잘 작성해 주셨어요! 추가로 `info.error()` 로깅을 사용해서 에러 발생 지점을 추적할 수 있도록 하면 배포된 서버에서 에러 모니터링이 더 수월해지는 효과를 얻게 될 것 같아요. 😄

supsup-hae commented on src/main/java/cotato/backend/global/common/ApiResponse.java:
```diff
@@ -0,0 +1,39 @@
+    private final boolean success;
+    private final T data;
+    private final String message;
+    private final String path;
```
에러 응답을 위해서 path를 명시해 주신 거로 생각합니다! 
제가 생각하기에는 Swagger를 이용하다 보니 API 호출 경로가 헷갈릴 염려는 없기 때문에 성공 응답에서는 `path`를 명시하지 않고 `data` 만 표시해도 괜찮을 것 같아요. 하지만 여전히 에러 응답에는 필요한 구조이다 보니 별도의 `ErrorResponse` 을 만들어보는 건 어떨까요? 그리고 `GlobalExceptionHandler` 내부에서 `ErrorResponse`를 이용해서 응답을 구조화하면 더 좋은 접근이 될 것 같습니다! 😁

supsup-hae replied:
> [공통 응답 작성](https://velog.io/@j2noo/Spring-ApiResponse-%ED%86%B5%EC%9D%BC%EC%84%B1%EC%9E%88%EA%B2%8C-%EA%B5%AC%ED%98%84%ED%95%98%EA%B8%B0SuccessResponse-ErrorResponse)
- 참고하면 좋을 것 같은 내용 첨부해둘게요

supsup-hae commented on src/main/java/cotato/backend/global/error/CustomException.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.http.HttpStatus;
+
+@Getter
+public class CustomException extends RuntimeException {
```
👏🏻

supsup-hae reviewed (changes requested):
수고하셨습니다! 리뷰 남겼으니 확인해 주시고 검토해 주시면 될 것 같습니다! 언제든 코멘트에 궁금한 점이 있다면 알려주세요.

supsup-hae commented on src/main/java/cotato/backend/global/common/ApiResponse.java:
```diff
@@ -0,0 +1,39 @@
+
+@Getter
+@AllArgsConstructor
+@Builder
```
```suggestion
@Builder
@JsonInclude(JsonInclude.Include.NON_NULL) 
```
해당 어노테이션을 사용하면 null 값은 응답에 포함하지 않게 구성이 가능합니다! 추가해주는것도 좋을 것 같아요!

jwdebbie replied:
피드백 주신 내용 반영 완료했습니다! 
다시 한 번 확인 부탁드립니다 :)

jwdebbie committed:
♻️ refactor: SuccessResponse/ErrorResponse 분리 및 예외 응답 구조 수정 (#3)

supsup-hae committed:
:recycle: refactor: ApiResponse 클래스 수정 (#3)

- ApiResponse 클래스의 구조를 개선하여 응답 데이터의 일관성과 재사용성을 높였습니다.

supsup-hae committed:
:recycle: refactor: ErrorResponse 클래스 수정 (#3)

- ErrorResponse 클래스의 구조를 개선하여 코드 가독성과 유지보수성을 높였습니다.
- 중복되는 필드 및 메서드를 정리하고, 예외 처리 시 응답 데이터의 일관성을 확보했습니다.

supsup-hae committed:
:recycle: refactor: GlobalExceptionHandler 예외 처리 개선 (#3)

- ErrorResponse 및 ApiResponse 구조를 활용해 예외 발생 시 클라이언트에 전달되는 메시지의 일관성을 유지했습니다.

supsup-hae committed:
:recycle: refactor: GlobalExceptionHandler 로깅 수정 (#3)
